### PR TITLE
[Fix] Fix demo and visualization

### PR DESCRIPTION
**Pull request by JingweiZhang12**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -82,7 +82,7 @@ a part of the OpenMMLab project developed by [MMLab](http://mmlab.ie.cuhk.edu.hk
 - **Support indoor/outdoor 3D detection out of box**
 
   It directly supports popular indoor and outdoor 3D detection datasets, including ScanNet, SUNRGB-D, Waymo, nuScenes, Lyft, and KITTI.
-  For nuScenes dataset, we also support [nuImages dataset](https://github.com/open-mmlab/mmdetection3d/tree/latest/configs/nuimages).
+  For nuScenes dataset, we also support [nuImages dataset](https://github.com/open-mmlab/mmdetection3d/tree/main/configs/nuimages).
```
Comment (Xiangxu-0103) [inline README.md:85]: Shall we mind the user that the refactor of `nuImages dataset` is still on the going?

File: configs/pointnet2/README.md [hunk 2]
```diff
@@ -35,7 +35,7 @@ We implement PointNet++ and provide the result and checkpoints on ScanNet and S3
 
 - Since ScanNet dataset doesn't provide ground-truth labels for the test set, users can only evaluate test set performance by submitting to its online benchmark [website](http://kaldir.vc.in.tum.de/scannet_benchmark/). However, users are only allowed to submit once every two weeks. Therefore, we currently report val set mIoU. Test set performance may be added in the future.
 
-- To generate submission file for ScanNet online benchmark, you need to modify the ScanNet dataset's [config](https://github.com/open-mmlab/mmdetection3d/blob/master/configs/_base_/datasets/scannet_seg-3d-20class.py#L126). Change `ann_file=data_root + 'scannet_infos_val.pkl'` to `ann_file=data_root + 'scannet_infos_test.pkl'`, and then simply run:
+- To generate submission file for ScanNet online benchmark, you need to modify the ScanNet dataset's [config](https://github.com/open-mmlab/mmdetection3d/blob/main/configs/_base_/datasets/scannet_seg-3d-20class.py#L126). Change `ann_file=data_root + 'scannet_infos_val.pkl'` to `ann_file=data_root + 'scannet_infos_test.pkl'`, and then simply run:
```
Comment (Xiangxu-0103) [inline configs/pointnet2/README.md:38]: scannet_seg-3d-20class.py#L126 -> scannet-seg.py#L139

File: docs/en/advanced_guides/datasets/scannet.md [hunk 3]
```diff
@@ -337,7 +337,7 @@ train_pipeline = [
 ## Testing and Making a Submission
 
 By default, our codebase evaluates semantic segmentation results on the validation set.
-If you would like to test the model performance on the online benchmark, add `--format-only` flag in the evaluation script and change `ann_file=data_root + 'scannet_infos_val.pkl'` to `ann_file=data_root + 'scannet_infos_test.pkl'` in the ScanNet dataset's [config](https://github.com/open-mmlab/mmdetection3d/blob/master/configs/_base_/datasets/scannet_seg-3d-20class.py#L126). Remember to specify the `txt_prefix` as the directory to save the testing results.
+If you would like to test the model performance on the online benchmark, add `--format-only` flag in the evaluation script and change `ann_file=data_root + 'scannet_infos_val.pkl'` to `ann_file=data_root + 'scannet_infos_test.pkl'` in the ScanNet dataset's [config](https://github.com/open-mmlab/mmdetection3d/blob/main/configs/_base_/datasets/scannet_seg-3d-20class.py#L126). Remember to specify the `txt_prefix` as the directory to save the testing results.
```
Comment (Xiangxu-0103) [inline docs/en/advanced_guides/datasets/scannet.md:340]: scannet_seg-3d-20class.py#L126 -> scannet-seg.py#L139

File: docs/en/notes/compatibility.md [hunk 4]
```diff
@@ -186,7 +186,7 @@ We adopt new pre-processing and conversion steps of ScanNet dataset. In previous
 
 - There is an `PointSample` pipeline in the data pipelines for ScanNet detection task which down-samples points. So removing down-sampling in data generation will not affect the code.
 
-We have trained a [VoteNet](https://github.com/open-mmlab/mmdetection3d/blob/master/configs/votenet/votenet_8x8_scannet-3d-18class.py) model on the newly processed ScanNet dataset and get similar benchmark results. In order to prepare ScanNet data for both detection and segmentation tasks, please re-run the new pre-processing scripts following the ScanNet [README.md](https://github.com/open-mmlab/mmdetection3d/blob/master/data/scannet/README.md/).
+We have trained a [VoteNet](https://github.com/open-mmlab/mmdetection3d/blob/main/configs/votenet/votenet_8x8_scannet-3d-18class.py) model on the newly processed ScanNet dataset and get similar benchmark results. In order to prepare ScanNet data for both detection and segmentation tasks, please re-run the new pre-processing scripts following the ScanNet [README.md](https://github.com/open-mmlab/mmdetection3d/blob/master/data/scannet/README.md/).
```
Comment (Xiangxu-0103) [inline docs/en/notes/compatibility.md:189]: votenet_8x8_scannet-3d-18class -> votenet_8xb8_scannet-3d.py

https://github.com/open-mmlab/mmdetection3d/blob/master/data/scannet/README.md/ -> https://github.com/open-mmlab/mmdetection3d/blob/main/data/scannet/README.md/

File: docs/en/notes/compatibility.md [hunk 5]
```diff
@@ -204,4 +204,4 @@ Please refer to the SUNRGBD [README.md](https://github.com/open-mmlab/mmdetectio
...
+In MMDetection 0.6.0, we updated the model structures of VoteNet and H3DNet, therefore model checkpoints generated by MMDetection \< 0.6.0 should be first converted to a format compatible with the latest structures via [convert_votenet_checkpoints.py](https://github.com/open-mmlab/mmdetection3d/blob/master/tools/model_converters/convert_votenet_checkpoints.py) and [convert_h3dnet_checkpoints.py](https://github.com/open-mmlab/mmdetection3d/blob/master/tools/model_converters/convert_h3dnet_checkpoints.py) . For more details, please refer to the VoteNet [README.md](https://github.com/open-mmlab/mmdetection3d/tree/main/configs/votenet/README.md/) and H3DNet [README.md](https://github.com/open-mmlab/mmdetection3d/tree/main/configs/h3dnet/README.md/).
```
Comment (Xiangxu-0103) [inline docs/en/notes/compatibility.md:207]:
```suggestion
In MMDetection 0.6.0, we updated the model structures of VoteNet and H3DNet, therefore model checkpoints generated by MMDetection \< 0.6.0 should be first converted to a format compatible with the latest structures via [convert_votenet_checkpoints.py](https://github.com/open-mmlab/mmdetection3d/blob/main/tools/model_converters/convert_votenet_checkpoints.py) and [convert_h3dnet_checkpoints.py](https://github.com/open-mmlab/mmdetection3d/blob/main/tools/model_converters/convert_h3dnet_checkpoints.py) . For more details, please refer to the VoteNet [README.md](https://github.com/open-mmlab/mmdetection3d/tree/main/configs/votenet/README.md/) and H3DNet [README.md](https://github.com/open-mmlab/mmdetection3d/tree/main/configs/h3dnet/README.md/).
```

File: docs/zh_cn/advanced_guides/supported_tasks/lidar_sem_seg3d.md [hunk 6]
```diff
@@ -71,7 +71,7 @@ mmdetection3d
 ## 测试与提交
 
 如果你只想在在线基准上进行推理或测试模型性能，你需要在配置文件中的 `test_evalutor` 字段增加 `submission_prefix`， 例如配置文件增加 `test_evaluator = dict(type='SegMetric',submission_prefix=work_dirs/pointnet2_ssg/test_submission`)。
-并将 ScanNet 数据集[配置文件](https://github.com/open-mmlab/mmdetection3d/blob/master/configs/_base_/datasets/scannet_seg-3d-20class.py#L129)中的 `ann_file=scannet_infos_val.pkl` 变成 `ann_file=scannet_infos_test.pkl`。在生成结果后，你可以压缩文件夹并上传至 [ScanNet 评估服务器](http://kaldir.vc.in.tum.de/scannet_benchmark/semantic_label_3d)上。
+并将 ScanNet 数据集[配置文件](https://github.com/open-mmlab/mmdetection3d/blob/main/configs/_base_/datasets/scannet_seg-3d-20class.py#L129)中的 `ann_file=scannet_infos_val.pkl` 变成 `ann_file=scannet_infos_test.pkl`。在生成结果后，你可以压缩文件夹并上传至 [ScanNet 评估服务器](http://kaldir.vc.in.tum.de/scannet_benchmark/semantic_label_3d)上。
```
Comment (Xiangxu-0103) [inline docs/zh_cn/advanced_guides/supported_tasks/lidar_sem_seg3d.md:74]: scannet_seg-3d-20class.py#L129 -> scannet-seg.py#L139

File: docs/en/notes/compatibility.md (hunk 4 above)
Comment (JingweiZhang12) [reply docs/en/notes/compatibility.md:189]: This is the description of version 0.14. We should not have modified `master` to `main`. I revert it.

File: docs/en/notes/compatibility.md (hunk 5 above)
Comment (JingweiZhang12) [reply docs/en/notes/compatibility.md:207]: The same as the above.

**Review by sunjiahao1999 — APPROVED** [review]

File: docs/zh_cn/model_zoo.md [hunk 7]
```diff
@@ -10,104 +10,104 @@
...
 
 ### FCAF3D
 
-请参考 [FCAF3D](https://github.com/open-mmlab/mmdetection3d/blob/master/configs/fcaf3d) 获取更多的细节，我们在 ScanNet, S3DIS 和 SUN RGB-D 数据集上给出了相应的基准结果。
+请参考 [FCAF3D](https://github.com/open-mmlab/mmdetection3d/blob/main/configs/fcaf3d) 获取更多的细节，我们在 ScanNet, S3DIS 和 SUN RGB-D 数据集上给出了相应的基准结果。
 
 ### Mixed Precision (FP16) Training
 
-细节请参考 [Mixed Precision (FP16) Training 在 PointPillars 训练的样例](https://github.com/open-mmlab/mmdetection3d/blob/master/configs/pointpillars/hv_pointpillars_fpn_sbn-all_fp16_2x8_2x_nus-3d.py)。
+细节请参考 [Mixed Precision (FP16) Training 在 PointPillars 训练的样例](https://github.com/open-mmlab/mmdetection3d/blob/main/configs/pointpillars/hv_pointpillars_fpn_sbn-all_fp16_2x8_2x_nus-3d.py)。
```
Comment (Xiangxu-0103) [inline docs/zh_cn/model_zoo.md:113]: hv_pointpillars_fpn_sbn-all_fp16_2x8_2x_nus-3d.py -> pointpillars_hv_fpn_sbn-all_8xb2-amp-2x_nus-3d.py. And the corresponding en doc.

File: mmdet3d/engine/hooks/visualization_hook.py [hunk 8]
```diff
@@ -97,13 +97,18 @@ def after_val_iter(self, runner: Runner, batch_idx: int, data_batch: dict,
...
+                'mono_det', 'multi-view_det', 'multi-modality_det'
+        ]:
+            assert 'img_path' in outputs[0], 'img_path is not in outputs[0]'
             img_path = outputs[0].img_path
             img_bytes = get(img_path, backend_args=self.backend_args)
             img = mmcv.imfrombytes(img_bytes, channel_order='rgb')
             data_input['img'] = img
 
-        if 'lidar_path' in outputs[0]:
+        if self.vis_task in ['lidar_det', 'multi-modality_det']:
```
Comment (Xiangxu-0103) [inline mmdet3d/engine/hooks/visualization_hook.py:109]: Whether need to add `lidar_seg`?

File: mmdet3d/engine/hooks/visualization_hook.py [hunk 9]
```diff
@@ -145,24 +150,37 @@ def after_test_iter(self, runner: Runner, batch_idx: int, data_batch: dict,
...
                 img_path = data_sample.img_path
                 img_bytes = get(img_path, backend_args=self.backend_args)
                 img = mmcv.imfrombytes(img_bytes, channel_order='rgb')
                 data_input['img'] = img
+                if self.test_out_dir is not None:
+                    out_file = osp.basename(img_path)
+                    out_file = osp.join(self.test_out_dir, out_file)
 
-            if 'lidar_path' in data_sample:
+            if self.vis_task in ['lidar_det', 'multi-modality_det']:
```
Comment (Xiangxu-0103) [inline mmdet3d/engine/hooks/visualization_hook.py:171]: The same as above.

**Pull request merged**